ENH:MAINT: sparse.linalg: rewrite ARPACK in C

ilayn opened the pull request:
#### Reference issue
Towards #18566

#### What does this implement/fix?
ARPACK in C as a static (or shared lib with BLAS/LAPACK dependency, it's complicated...) library.

#### Additional information

 - I included a hand-written version of the notorious `(cz)dotc` BLAS functions so hopefully we will never hear from them.
 - Now ARPACK is thread-safe and reentrant since all SAVE'd variables are taken out to the reverse communication layer with a struct. Hence closed but still needs testing against #5846 EDIT: **Checked and verified**
 - Some of the functions are very confusing and perform Dijkstra-raging goto tricks such as `?NAPPS.f` so I partially rewrote them instead of translating, because each time another week was being spent in debugging if something was not right.
 - Now the pseudo-random generation happens outside ARPACK hence `rng` keyword is added. To be able to use it, new job codes `ido = 4` and `ido = 5` are introduced and `ido = -1` is removed.
 - Since now ARPACK is thread-safe/reentrant, those tests that are testing for "no crash" behavior are removed/let passed.
 - LAPACK `dlamch('E')` by default resolves to `0.5*eps`, in this PR `eps` is used for tolerance tests.

#### All open issues regarding ARPACK

 - [ ] #3716 
 - [ ] #6827 
 - [ ] #9284 
 - [ ] #11232 
 - [ ] #11261  ILP64
 - [ ] #13033 Add the picture
 - [ ] #17632 LinearOperator support
 - [ ] #19235 Performance
 - [ ] #20808 More hanging tests

Closes #9023 
Closes #9185
Closes #9460
Closes #10744
Closes #11805 
Closes #13222 
Closes #13692 
Closes #13780 
Closes #14823 
Closes #18191 
Closes #18736
Closes #19756 
Closes #20339
Closes #21137 
Closes #22310

ilayn committed:
MAINT:sparse.linalg: Remove ARPACK F77 code

ilayn committed:
ENH:sparse.linalg: Add ARPACK C code for nonsym real

ilayn committed:
ENH:sparse:linalg: Add ARPACK header file

ilayn committed:
ENH:sparse.linalg: Add ARPACK C code for nonsym complex

ilayn committed:
MAINT:sparse.linalg: Add MSVC guards for complex definitions

SQ: remove complexdefs.h

ilayn committed:
ENH:sparse.linalg: Add ARPACK C code for sym real

ilayn committed:
MAINT:sparse.linalg: Adjust meson file for ARPACK C rewrite

ilayn committed:
MAINT:sparse.linalg: Adjust arpack.py for C rewrite

[skip ci]

ilayn committed:
MAINT: sparse.linalg: Add undefs for custom macros to header

ilayn committed:
MAINT: sparse.linalg: Rename the ishift field to shift in nonsym

ilayn committed:
MAINT: sparse.linalg: Fix the LAPACK call syntax in sym code

ilayn committed:
MAINT: sparse.linalg: Finish ARPACK full compiled nonsymmetric real solver

ilayn committed:
MAINT: sparse.linalg: Finish ARPACK full compiled nonsym complex solvers

ilayn committed:
MAINT: sparse.linalg: Finish ARPACK full compiled sym real solvers

ilayn committed:
MAINT: sparse.linalg: Finish ARPACK full compiled header file

ilayn committed:
MAINT: sparse.linalg: Finish ARPACK full compiled meson file

[skip ci]

ilayn committed:
MAINT: sparse.linalg: Modify ARPACK Python driver file

[skip ci]

ilayn committed:
MAINT: sparse.linalg: Clean up gitignore file for removed F77 code

[skip ci]

ilayn committed:
MAINT: sparse.linalg: Add new module file for ARPACK interface

ilayn committed:
MAINT: sparse.linalg: Remove the ARPACK int selection

ilayn committed:
MAINT: sparse.linalg: Adjustments to the arpack.py file after rewrite

ilayn committed:
MAINT: sparse.linalg: Add nonsymmetric symbols to the header file

ilayn committed:
MAINT: sparse.linalg: ARPACK nonsym real fixes and typos

ilayn committed:
MAINT: sparse.linalg: Remove dependence to complex dotc routines in ARPACK

ilayn committed:
MAINT: sparse.linalg: Create the ARPACK extension module code

ilayn committed:
MAINT: sparse.linalg: Adjust meson file for ARPACK ext module

[skip ci]

ilayn committed:
MAINT: sparse.linalg: Adjustments to the arpack.py file for rng

ilayn committed:
MAINT: sparse.linalg: Temporary changes to module for testing

ilayn committed:
MAINT: sparse.linalg: Add ido=5 for random OPX ARPACK operation

ilayn committed:
MAINT: sparse.linalg: Fix comments in nonsymmetric complex code

ilayn committed:
MAINT: sparse.linalg: Printed version of nonsym real code for segfault fixes

[skip ci]

ilayn committed:
MAINT:sparse.linalg: Add more debugging to nonsym ARPACK code

ilayn committed:
ENH:sparse.linalg: Add sneupd to the ARPACK module

[skip ci]

ilayn committed:
MAINT:sparse.linalg: More troubleshooting segfaults

[skip ci]

ilayn committed:
MAINT: sparse.linalg: Segfaults fixed in ARPACK nonsym real solver

[skip ci]

ilayn committed:
MAINT:sparse.linalg: Fixes various nonsym real bugs verified large examples

[skip ci]

ilayn committed:
FIX:sparse.linalg: Typo in the post-processing nonsym real eigs

[skip ci]

ilayn committed:
ENH:sparse.linalg: Complex nonsym solver first pass

[skip ci]

ilayn committed:
MAINT:sparse.linalg: Add zneupd to module and convert to multi-phase init

[skip ci]

ilayn committed:
MAINT:sparse.linalg: Add missing casts for NumPy pointers in ARPACK module

[skip ci]

ilayn committed:
FIX:sparse.linalg: Modify ARPACK codes for random array on Python side

[skip ci]

ilayn committed:
ENH:sparse.linalg: Add nonsym complex protos to header file

[skip ci]

ilayn committed:
ENH:sparse.linalg: Add nonsym complex ARPACK solvers

[skip ci]

ilayn committed:
TMP:sparse.linalg: Turn on pedantic flag on ARPACK

[skip ci]

ilayn committed:
SQ: Arpack.py

ilayn committed:
ENH:sparse.linalg: Complete the rewrite of ARPACK sym solvers

[skip ci]

ilayn committed:
ENH:sparse.linalg: Complete all the function wrapping in ARPACK module

[skip ci]

ilayn committed:
SQ: arpack.py

ilayn committed:
ENH:sparse.linalg: Expose all the functions in ARPAKC header

ilayn committed:
FIX:sparse.linalg: Allow for user defined initial vector in NonSym real ARPACK

ilayn committed:
TST:sparse.linalg: Allow for the reentrant ARPACK test

[skip ci]

ilayn commented:
Took a while but now ready for review. Nonsymmetric tests locally pass. Symmetric solver has a minor issue that I will follow-up when I have a bit more time.

dschmitz89 commented:
This is the largest diff I've ever seen. Any suggestions how to review? ;)

From top of my head: are there tests for the new numpy rng seeding mechanism?

ilayn commented:
I have them, currently the symmetric part is broken so I didn't send those parts. There is really no diff on the low-level code as it is old out, new in. I will have some time in the weekend to fix and we can start pounding with test problems, small or large.

ilayn committed:
BUG:sparse.linalg: Fix the evec reordering in nonsym real ARPACK solver

ilayn committed:
BUG:sparse.linalg: Fix reordering in nonsym complex ARPACK solver

ilayn committed:
BUG:sparse.linalg: Clear error code after aupd calls in ARPACK

szhorvat commented:
Are you planning to keep this API-compatible with ARPACK? It does not seem to be compatible at the moment.

In igraph, we provide an option to either link to an external ARPACK, or use the vendored one (to make compilation simple, including in Windows). My question is whether we can continue to do the same if using this C translation.

BLAS and LAPACK have at this point become de-fact _interfaces_ instead of libraries. There are multiple implementations in multiple programming languages. In principle, the same could happen with ARPACK.

Better yet, a long-term goal could be to rewrite ARPACK-NG in C and maintain the C version going forward. Have you spoken with the ARPACK-NG maintainers? My impression was that the lack of Fortran expertise has been an issue for them as well, so they might actually prefer doing this.  But it would only work if API-compatibility is maintained.

ilayn commented:
> Are you planning to keep this API-compatible with ARPACK? It does not seem to be compatible at the moment.

Yes, the API has changed, or broken, depending on how you look at it. there are many issues that can't be solved without changing the API and nobody wants to call functions with 23 arguments. There is no way to control the determinism with its quirky random number generator, somehow mysteriously segfaults which nobody can figure out why, F77 building requires bunch of nasty tricks and warning silencing and so on. Hence this is a hail-mary attempt to get rid of this maintenance burden. 

That's why I decided to reach out to you folks and others, in case folks also want to move on. But I can understand this is not everyone's favorite move. 

> In principle, the same could happen with ARPACK.

I don't believe that ARPACK or even LAPACK should have been in this state for the last decades. After 30 years, it is very difficult to change things indeed but every time this argument is invoked, we are poisoning the future to save the past. The very existence of ARPACK-NG is a proof of that. There have been a lot of attempts or independent libraries; a handful new ?PACKs, Spectra, Feast, many Julia libraries and so on. All of them somehow start from scratch but do not reach to the feature complete state of ARPACK for some reason. I am aware that I am in the minority in this line of thinking, but I don't think there is any other way. 

My goal here is to provide the functionality to SciPy without the F77 luggage, close as many open issues as possible, but not really maintain ARPACK further. There are about 800+ open issues mentioning ARPACK on Github. I don't have any appetite to be the attention center of that much trouble. 

> Have you spoken with the ARPACK-NG maintainers? My impression was that the lack of Fortran expertise has been an issue for them as well, so they might actually prefer doing this. But it would only work if API-compatibility is maintained.

No I have not. But in the past, we have also similar interactions with them and they also made it very clear that they don't want to touch the code but make sure that systems keep building and packaging it. I can't blame them. Even though I know exactly what ARPACK is doing (at research level), I still get tripped up in the go to mess and spend days staring at it. 

I think you are exactly on point that we probably should not call this ARPACK, but rename it and move on.

ilayn commented on scipy/sparse/linalg/_eigen/arpack/ARPACK/_arpack_n_double.c:
```diff
@@ -0,0 +1,2116 @@
+static int sortc_LI(const double, const double, const double, const double);
+static int sortc_SI(const double, const double, const double, const double);
+
+static const double unfl = 2.2250738585072014e-308;
+// static const double ovfl = 1.0 / 2.2250738585072014e-308;
+static const double ulp = 2.220446049250313e-16;
```
No, not really but let me change it anyways.

ilayn committed:
MAINT:sparse.linalg: Remove unused variables from nonsym real ARPACK code

ilayn committed:
MAINT:sparse.linalg: Remove unused variables from nonsym complex ARPACK code

ilayn committed:
MAINT:sparse.linalg: Remove hangups from sym ARPACK solver

ilayn committed:
MAINT:sparse.linalg: Remove compiler flags from ARPACK meson file

[skip ci]

yixuan commented:
Hi, this is the author of [Spectra](https://github.com/yixuan/spectra).

I came from https://github.com/igraph/igraph/issues/551 and just realized that there are several scientific computing projects considering the re-implementation of ARPACK, so I try to share some of my experience regarding this issue.

One of the motivations of starting the Spectra project is exactly what @ilayn and @szhorvat encounter today: ARPACK was written in FORTRAN 77 code, which causes various compiling warnings/errors on modern building systems. Moreover, the API of ARPACK is very "classic", and using ARPACK is not as simple as calling a function -- it requires you to organize your whole program according to the structure of the algorithm: you are responsible of initializing the problem, implementing the linear operators, and responding to the return values of the ARPACK functions, etc. All of these designs are okay in the time when ARPACK was created, but it is hardly a pleasant experience to include ARPACK into your own project nowadays.

So Spectra is a complete rewriting of ARPACK, or more precisely, a re-implementation of the algorithm that ARPACK is based on. The current status is that all solvers in ARPACK have been realized except for the non-Hermitian complex eigen solvers. Yes, it depends on the Eigen library, which seems to be the main reason that igraph and SciPy cannot directly use it. Let me just explain it a little bit. Eigen is useful to me since it is convenient to manage matrix objects, and it is header-only, meaning that I even do not need to depend on BLAS and LAPACK libraries. But the essential part of Spectra is somewhat independent of Eigen: the internal implementation still heavily uses pointers, and I can replace matrix-related functions by BLAS/LAPACK calls.

If there is indeed a demand, I would be interested in creating a BLAS/LAPACK version of Spectra, but I want to understand whether this is what you want. Could you kindly comment on the following questions if interested?

1. Do you allow C++ code, and if yes, in what standard?
2. Do you need a completely API-compatible alternative to ARPACK? If this is the case, then probably the only choice is to do a language-level F2C translation.
3. For matrix operations, do you prefer calling BLAS/LAPACK as much as possible, or prefer minimizing external libraries as much as possible?
4. Do you need the computation result to be exactly equal to the ARPACK version? In my own experience, this is the hardest part, since different implementations may have different rounding and convergence properties. ARPACK is based on an iterative algorithm, so various errors may be accumulated during the iterations. ARPACK also has some cryptic tricks in the code that are hard to interpret.
5. Apart from the Eigen dependence, how do you feel about Spectra's API? Does it fit your potential need?

I am trying to collect this information to see whether it is of value to make a version of Spectra that is independent of Eigen. In fact, I have done this once before, when I used Spectra-derived code (e.g., the `newarp_*` files in [here](https://gitlab.com/conradsnicta/armadillo-code/-/tree/14.4.x/include/armadillo_bits)) to implement the `eigs_sym()` and `eigs_gen()` functions in [Armadillo](https://gitlab.com/conradsnicta/armadillo-code).

ilayn commented:
Hi @yixuan first of all, on behalf of all community, a true congratulations from me for attacking this problem in such a comprehensive fashion. A very commendable work indeed. 

First, I should clarify why I have done this work; it is to clean up the F77 from SciPy so that we resolve a lot of decades long issues all in one go. Hence it is slightly different picture when it comes to our motivation. It is getting significantly difficult to maintain this codebase while the rest of the ecosystem is moving to better ergonomics and tooling. Hence it is one of many items we have tackled all listed in #18566. This does not take away any great value ARPACK provided thanklessly over the decades and still do. However the way it provides the value has been a continuous source of pain. Hence this work. 

Regarding your questions, let me first get them all at once and I can comment at the end;

 1. Yes we do allow, and in fact we have Highs and unuran libraries already. Our standard is C++17.  
 2. No I already broke the API by collecting all `SAVE`d variables into a struct and pass it back and forth for reverse communication layer. You can see the definition in the `_arpack.h` file. This allows us to use ARPACK in thread-safe envs or in re-entrant use cases. I am not sure if I understand if you tackled this or not since I don't know the new names of the variables. 
 3. Yes, we only use BLAS/LAPACK interface. The backend whether OpenBLAS/MKL/Blis etc. can be swapped. We don't want to have homebaked versions of the linalg routines as they typically rot in a decade as LAPACK moves on.
 4. Bitwise equivalence is not sought after. I think that is an unreasonable requirement, but if results matches to a reasonable degree we consider it acceptable. Precise definition of reasonable is unfortunately on a case-by-case bases. 
 5. I think the API looks good and clean. Nice work there. For us we don't mind too much about the potential API as it is going to be buried behind the Python layer.

A quick comment on the API compatible ARPACK. Though it sounds like much, the work here is an independent rewrite of ARPACK, mostly mechanical line-by-line translation (as f2c would do) but rewritten partially where the codebase does some unhinged stuff predominantly `??apps.f` and the `?stqrb.f`. So it is pretty much a painstaking translation (SPAG or LLMs did not help one bit unfortunately against this codebase). Whether an API compatible ARPACK is needed I can't judge however I think we should stop using the F77 code regardless of its convenience. 

So because the save variables are now in a struct, you'll see a lot of `V->xxx` in the code which is the struct that the caller sends in to preserve the state. If needed those can be re-expanded by regex + copy pasta and you can regain the API compatibility. Note that you only need to do this for the exposed functions hence 8 function in total. That's quite doable. Hence if you need compatible ARPACK it can be done but then you need to add static keywords to preserve the statefulness to all relevant variables. It's not complicated or tedious but it's work.

I didn't because I am trying to solve the thread-safety problem and not try to offer a drop-in replacement for ARPACK hence the breakage. But today I'll fix the last bugs in the symmetric solver and then we have an all green tested and feature complete working ARPACK.

I am still trying to guess if a rename is going to be better or complicate the matters.

ev-br commented:
A partial and opinionated wishlist:

> Do you allow C++ code, and if yes, in what standard?

SciPy allows C++17, cf https://scipy.github.io/devdocs/dev/toolchain.html

> Do you need a completely API-compatible alternative to ARPACK? If this is the case, then probably the only choice is to do a language-level F2C translation.

No. We would welcome an API which takes the best of ARPACK and leaves the rest of ARPACK :-).

> For matrix operations, do you prefer calling BLAS/LAPACK as much as possible, or prefer minimizing external libraries as much as possible?

We're talking about a single dependency on a BLAS/LAPACK provider, I think. A tricky part, for SciPy at least,  is that we allow several of those: OpenBLAS, MKL, BLIS, Accelerate, ATLAS etc. While the API is fixed, ABI and build details differ quite a bit between them. 

In an ideal world, dense linear algebra is tucked into a mix-and-match layer, so that one user library can build with Eigen and no LAPACK, SciPy can use its LAPACK-selection logic, scipy redistributors (distros, for example), use their logic, and so on. We will also want to be able to switch between 32-bit integer and 64-bit integer variants (LP64 vs ILP64) at build time.

I fully realize this asks for a substantial bump in complexity from a header-only library! 

> Do you need the computation result to be exactly equal to the ARPACK version? ... ARPACK also has some cryptic tricks in the code that are hard to interpret.
 
Exact equivalence is probably not achievable and certainly not sought after.
The _cryptic tricks_ would be great to enumerate/document somewhere. Some of these are likely artifacts from the era; some might amount to a chapter of somebody's PhD thesis some time ago, and could be still useful to take along. Having some pointers would be of immense help---even if investigating a these cryptic tricks will amount to a PhD thesis chapter tomorrow or the day after.

ev-br commented:
> I am still trying to guess if a rename is going to be better or complicate the matters.

The former, I'd say. It's not exactly compatible, so a rename would be helpful. Together with a README or something, listing basically your comment Ilhan, https://github.com/scipy/scipy/pull/22748#issuecomment-2953897553

szhorvat commented:
Thanks for the clarification @ilayn and sorry about the misunderstandings!

> One is the randomization which is quite arcane, and two it can't be re-entrant.

These are issues for us as well.

> Let me get this in a proper completed shape so you can start testing the results for yourself and decide using SciPy interface.

Will definitely try it out.

ilayn commented:
Let me provide some concrete details what we mean by API breakage, take non-symmetric double precision purifier `dneupd`

```Fortran
      subroutine dneupd (rvec , howmny, select, dr    , di,
     &                   z    , ldz   , sigmar, sigmai, workev,
     &                   bmat , n     , which , nev   , tol,
     &                   resid, ncv   , v     , ldv   , iparam,
     &                   ipntr, workd , workl , lworkl, info)
```

Instead of this now the equivalent call is 

```C++
dneupd(struct ARPACK_arnoldi_update_vars_d *V, int rvec, int howmny, int* select,
       double* dr, double* di, double* z, int ldz, double sigmar, double sigmai,
       double* workev, double* resid, double* v, int ldv, int* ipntr, double* workd,
       double* workl)
```

It's not too different since most of these are array arguments and can't be tucked in anywhere but the complete state is now hidden inside that initial V, if I may compress the fields

```C++
struct ARPACK_arnoldi_update_vars_d {
    double tol,getv0_rnorm0, aitr_betaj, aitr_rnorm1, aitr_wnorm, aup2_rnorm;
    enum ARPACK_ido ido;    
    enum ARPACK_which which, bmat, info, iter, maxiter;            
    int mode, n, nconv, ncv, nev, np, numop, numpb, numreo, shift, getv0_first, getv0_iter, getv0_itry, getv0_orth, aitr_iter, aitr_j, aitr_orth1, aitr_orth2, aitr_restart, aitr_step3, aitr_step4, aitr_ierr, aup2_initv, aup2_iter, aup2_getv0, aup2_cnorm, aup2_kplusp, aup2_nev0, aup2_np0, aup2_numcnv, aup2_update, aup2_ushift;
};
```

You can see from the prefixes which function they are `SAVE`d. These are the ones that are kept on the reverse communication layer to keep the state intact, a basic common trick that game devs use all the time for characters. So it's not really anything that has changed from ARPACK but the way we interact with it is changed. Whether this is justified or not is a different story.

yixuan commented:
Thanks for the discussions! Now I understand the current status better, and it seems that several things can be done in parallel. This PR for SciPy looks to be in a good shape, and I will also try to use my spare time to draft a LAPACK-based version of Spectra. It is always nice to have more options in open source projects. 😁

When I get something that is more concrete, I'll let you know.

szhorvat commented:
@ilayn What values are you using for `bmat` instead of characters `I` and `G`? Which is true and which is false, out of `I` and `G`?

Given that you are treating this as a boolean, is there any reason you're not using the C99 `bool` type? Also, sice this is treated as a boolean, it is perhaps better not to validate the value and strictly enforce a `0` or `1` when I read "boolean", I assume that values will only be tested using conditionals, not `==`, and any nonzero is treated as "true".

ilayn commented:
Yes that's fair. I was being lazy.

szhorvat commented:
Some more comments:

----

To make porting from classic ARPACK easier:

 - Would it be possible to annotate the entries of the `ARPACK_arnoldi_update_vars` struct with whether they are used for input, output or both? A comment in the header would do. For example, `ido` is for both input and output. `ncv` is input only. I don't remember if there are output-only ones. Of course this can be mostly looked up in the original ARPACK docs.
 - Similarly, it'd be nice to make notes about which struct members the caller is supposed to set vs which are private to ARPACK's internal operations.
 - It'd be nice to have some notes on how to port from classic ARPACK, in particular for function parameters that appear to have disappeared at first glance (I still haven't figured all of them out). Examples are IPNTR, IPARAM (broken down into pieces, which is good), LWORKL.

----

What do you think about prefixing most public symbols with `ARPACK`, similar to how `ARPACK_which` is prefixed? I am primarily thinking of enum values, such as `which_LM` -> `ARPACK_which_LM`. This is to guarantee no conflict with other packages.

To be clear, this is NOT a problem for igraph.

----

What are the `ARPACK_cplx` macros for? They seem to be unused.

----

As I recall, there were some issues related to specific Fortran calling conventions and string parameters. GFortran would require extra parameters to indicate the length of string (`character`) parameters. Unfortunately I don't have a full understanding of this issue. Is this not going to be a problem when calling LAPACK routines that have string parameters?

----

From a maintainability perspective, would it make sense to use some macro magic and share code between the double-precision and single-precision variants?

ilayn committed:
BUG:sparse.linalg: Fix a few bugs in ARPACK sym solvers

ilayn committed:
TST:sparse.linalg: Enable parallel and reentrance tests

ilayn committed:
TST:sparse.linalg: Parametrize ARPACK tests

ilayn commented:
Thanks @szhorvat good points overall. 

> Would it be possible to annotate the entries of the ARPACK_arnoldi_update_vars struct with whether they are used for input, output or both?

Yes, that's a good idea. 

> Similarly, it'd be nice to make notes about which struct members the caller is supposed to set

It's in my notes for the docs, I'll get to it once this symmetric bug is done.

> What do you think about prefixing most public symbols with ARPACK, similar to how ARPACK_which is prefixed?

Yeah, probably we should do that, we need a name for this first though. 

> These are actually used but not in the _arpack.h header.

Must be a relic from earlier commits. I'll remove it. 

> As I recall, there were some issues related to specific Fortran calling conventions and string parameters. 

I'm also not very involved in that one but we are linking to the BLAS/LAPACK libraries and not compiling together with ARPACK. So that problem lives in BLAS/LAPACK vendors in case they have it, we are just exposing the symbols.

> From a maintainability perspective, would it make sense to use some macro magic and share code between the double-precision and single-precision variants?

This is actually what I am doing though not in C. I have some simple regex file that I run through the copy of the double version to generate single versions. Preprocessors are text replacement ops anyways hence I don't think they are pretty arcane text processing tools. I have better ones so I don't use such macros in general. Instead of making the code unreadable I typically do it externally. But it should be possible indeed with macro magic. 

<hr>

I have one last symmetric bug somewhere. I roughly know where it is but I ran out of time that I'll be busy for a few days. If I can find it by staring, I'll fix that one otherwise I'll need to find more time to sit down for a fix. Then we can start the cosmetics and other changes.

ilayn committed:
MAINT:sparse.linalg: Remove Cpp guards from header

ilayn committed:
BUG:sparse.linalg: Correct double parsing in dseupd

ilayn committed:
TST:sparse.linalg: Remove unused import

ilayn committed:
BUG:sparse.linalg: Fix typos in ARPACK nonsym complex solver

ilayn committed:
MAINT:sparse.linalg: int guard ARPACK maxiter value

ilayn committed:
MAINT:sparse.linalg: Remove spurious entries from the ARPACK struct and rename symbols

ilayn committed:
MAINT:sparse.linalg: Remove spurious entries from the ARPACK dict

ilayn committed:
MAINT:sparse.linalg: Add ARPACK license, struct member info and rename symbols

ilayn committed:
MAINT:sparse.linalg: Rename exposed symbols with ARPACK prefix

ilayn committed:
MAINT:sparse.linalg: Rename exposed symbols with ARPACK prefix

ilayn committed:
MAINT:sparse.linalg: Rename exposed symbols with ARPACK prefix

ilayn commented:
> Please correct me if I misunderstood, you are at this stage waiting for SciPy-specific feedback, right?

Yes indeed. Then I'll start the cosmetics and document the API for external usage. 

> I might bombard you with some questions during the process.

No worries, that's good exercise anyways to improve ARPACK outside SciPy.

ilayn commented:
Ah yes, I basically forgot to add the readme and the license files. I'll add in the new PR I'll send out soon for restructuring. Since we decided to rename this lib anyways, due to API changes; ARPACKX, ARPAX or something like that. Everything else has been taken over the decades it seems.

For 2, that's a bit of a contested area, everybody has their own opinion about them because there is no structure dictated about the build discipline by the language. h files only go out of source directory if they are the public exposed symbols under an `include` directory and the internal header files are typically kept private. Depending on the size of the project either they are bundled with the src files or have an internal directory for .h files. But again these are all artificial conventions. If the project is small enough it does not matter. I converted it into `include`/`src` distinction and added a makefile in the upcoming PR. 

For 3, this is everyone's favorite item and I have been trying C++ workflow in the last few weeks with PROPACK. I can't say it gives me the "coding once, works twice" feeling. To be honest, tempita feels superior to template programming with C++. Because there is no continuous questioning of what will happen if I write constexpr or static const or whatever, am I falling into a UB, is this C++17 or newer etc. And debugging with C++ is more challenging. Instead coding once for double and running a regex over the file (converting `dgemm` to `sgemm` or `hypot` to `hypotf` etc.) is much much faster for me and diff'ble. 

I would not object to if someone else wants to combine the C sources and make a C++ lib with it but translating directly into C++ is really not my thing. Or let me say it differently, after I convert it to C, I don't have any opinion what happens to the code as it happened to `xsf`. I have no dog in that race :)

ilayn commented:
Ah OK,  I didn't know that. What a strange automation choice. But probably exists for a reason.

h-vetinari reviewed (commented):
Great work as always, though there are some implementation details that I find not ideal.

ilayn commented on scipy/sparse/linalg/_eigen/arpack/_arpackmodule.c:
```diff
@@ -0,0 +1,1078 @@
+#define PYERR(errobj,message) {PyErr_SetString(errobj,message); return NULL;}
+static PyObject* arpack_error;
+
+// The following macros are used to define the field names in the ARPACK struct.
+#define STRUCT_INEXACT_FIELD_NAMES X(tol) X(getv0_rnorm0) X(aitr_betaj) X(aitr_rnorm1) X(aitr_wnorm) X(aup2_rnorm)
+#define STRUCT_INT_FIELD_NAMES X(ido) X(which) X(bmat) X(info) X(iter) X(maxiter) X(mode) \
+                               X(n) X(nconv) X(ncv) X(nev) X(np) \
+                               X(shift) X(getv0_first) X(getv0_iter) X(getv0_itry) X(getv0_orth) \
+                               X(aitr_iter) X(aitr_j) X(aitr_orth1) X(aitr_orth2) X(aitr_restart) \
+                               X(aitr_step3) X(aitr_step4) X(aitr_ierr) X(aup2_initv) X(aup2_iter) \
+                               X(aup2_getv0) X(aup2_cnorm) X(aup2_kplusp) X(aup2_nev0) X(aup2_np0) \
+                               X(aup2_numcnv) X(aup2_update) X(aup2_ushift)
+#define STRUCT_FIELD_NAMES STRUCT_INT_FIELD_NAMES STRUCT_INEXACT_FIELD_NAMES
```
I commented on your PR for the reasoning on my part
